### ISSUE TITLE
Suggestion - devise module settings in initializer not model

### ISSUE THREAD

**Issue by LeKristapino** [plataformatec/devise#5147]
Has anybody brought up why devise modules that affect not only models, are placed in models not initializer?

It is weird that a line in model directly affects routes. Might it not be better to put modules in devise initializer so that instead of 
```
class User < ApplicationRecord
  devise :database_authenticatable, :validatable
end
```

it would be something like
```
Devise.setup do |config|
  config.devise_model User, :database_authenticable, :validatable
end
```

just food for thought

**Comment by colinross** [comment]
While I would admit the coupling is a bit akward as you noted, the catalyst is the scoping when using multiple models (resource types). The alternative would be having domain requirements (Users are recoverable but Admins are not, etc.) defined away from the Domain Model.

**Comment by LeKristapino** [comment]
To adress the scoping issue, I propose just model name as the first parameter( as in the example). We don't even need to change that much logic under the hood (as a first step), since it should be possible to inject functionality in models exactly the same way it is done now, just doing it behind the scenes.

**Comment by LeKristapino** [comment]
Another added bonus, is that this way it is possible to configure devise per-resource-type, as the configuration logic can be scoped to serve a specific resource (for example different unlock strategies, for User and Admin, etc.)

**Comment by colinross** [comment]
My understanding was that, it actually gets set up via the the `devise_for` call in the routes file (the right place). In turn, yes, it looks at the  model and detects which modules are loaded and then loads the appropriate (default) routes if not overrides are given.

That said, I think your premise of "a line in model directly affects routes" is incorrect. It is more specifically, a line in routes checks the model (which is configured in the model file) to create routes.

Simply put, the model definition does *not* create routes, but it informs the creation of routes depending on the modules loaded.

**Comment by tegon** [comment]
> Simply put, the model definition does not create routes, but it informs the creation of routes depending on the modules loaded.

That is correct.

> Another added bonus, is that this way it is possible to configure devise per-resource-type, as the configuration logic can be scoped to serve a specific resource (for example different unlock strategies, for User and Admin, etc.)

This would require a bigger change, right? I don't think that simply changing where the code gets injected in the model would also make this possible since some of the configurations are global.

**Comment by LeKristapino** [comment]
>> This would require a bigger change, right? I don't think that simply changing where the code gets injected in the model would also make this possible since some of the configurations are global.

Yes, but my suggestion to move the configuration from model to initializer just opens this possibility for future improvements. I do not suggest doing all of this in one go.

**Comment by tegon** [comment]
> Yes, but my suggestion to move the configuration from model to initializer just opens this possibility for future improvements. I do not suggest doing all of this in one go.

I do think the configuration per-resource could be done with the current design too. I understand what you saying about code organization but I'm struggling to see why your proposal would be more flexible than the way it is today.

> the method delete_all_data does not actually delete data, it just calls another method.

IMO, that's true. The method `delete_all_data` has a dependency on the class name `SomeModel` and its method `delete_all_data`. By dependency, I mean that this method would have to change whether `SomeModel` changes its name, the parameters or the name of its own `delete_all_data` method. But if `SomeModel` changes its implementation of `delete_all_data` while keeping the public API the same, then the parent method does not have to change at all.

**Comment by colinross** [comment]
The 'best practice' in rails-centric libraries is convention over configuration.
The convention (90% use case) is that if someone has a certain module enabled for their devise resource, `Recoverable` as an example, they also want to have the routes needed to make that work.

If you _really_ need to de-couple these concerns, you may do so at the model and routes level, via overrides to the  `devise_for, respectively.

The initializer **is** for global-level configuration (in this case global denotes applying to all the devise resources in use). If you want it to apply only to one resource-type, use options in the model or routes definition calls. You can disable creating routes altogether and create them manually if you wish as well.

I would also say as an aside, that it seems the concerns you have with de/coupling may have to do with an approach that doesn't see the library as a container [block box if you will] itself. The fact that devise informs itself through inference (via the model) rather than hard-coded configuration in the routes is just a choice for convenience of testing/implementation and keeping the library DRY. 

Do you have a specific use case / failing test case that you can't accomplish with the current form?

**Comment by LeKristapino** [comment]
No, it was just something that I thought would be a better way of configuring devise when looking for a problem why some devise routes weren't being generated. As I was continuously searching for something in the model itself, I realized that it seems somewhat odd to look for routes config in the model.

This was just to present a potential alternate approach in the configuration. There are no problems or cases that I've come across so far where this approach would fail to deliver. :)

**Comment by tegon** [comment]
> As I was continuously searching for something in the model itself, I realized that it seems somewhat odd to look for routes config in the model.

@LeKristapino That's a fair point, it would be easier to find out how things work. But this kind of change requires work both from maintainers and users of the Gem. We want to make it easier for the users to upgrade to new versions, that's why we'll leave as it is.

Thanks again for bringing this up and for the discussion.
